paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc

```diff
@@ -0,0 +1,518 @@
+    auto mask = scale.size() > 1 ? 1 : 0;
+    mkldnn::primitive_attr attr;
+    attr.set_output_scales(mask, scale);
+
+    auto src_mem =
+        memory({src_desc, this->engine_}, const_cast<void *>(src_data));
+    auto dst_mem = memory({dst_desc, this->engine_});
+
+    auto reorder_pd = mkldnn::reorder::primitive_desc(
+        src_mem.get_primitive_desc(), dst_mem.get_primitive_desc(), attr);
+
+    auto reorder = mkldnn::reorder(reorder_pd, src_mem, dst_mem);
+    stream(stream::kind::eager).submit({reorder}).wait();
+
+    return dst_mem;
+  }
+
+  memory TransposeInputY(const Tensor *input_y, memory::data_type type) {
+    auto dims = framework::vectorize2int(input_y->dims());
+    std::swap(dims[0], dims[1]);  // Correct output dimensions
+    auto src_desc = this->CreateMemDescriptor(dims, type, memory::format::io);
+    auto dst_desc = this->CreateMemDescriptor(dims, type, memory::format::oi);
+    return this->Reorder(src_desc, dst_desc, input_y->data<K>());
+  }
+
+  memory QuantInputY(memory input_y, const std::vector<float> &scale_y) {
+    const auto &dims = input_y.get_primitive_desc().desc().data.dims;
+    auto ndims = input_y.get_primitive_desc().desc().data.ndims;
+    auto y_dims = std::vector<int>(dims, dims + ndims);
+    auto y_dt = platform::MKLDNNGetDataType<K>();
+
+    auto y_desc = this->CreateMemDescriptor(y_dims, memory::data_type::s8,
+                                            memory::format::oi);
+
+    auto user_y_desc =
+        this->CreateMemDescriptor(y_dims, y_dt, memory::format::oi);
+
+    return ReorderWithScale(user_y_desc, y_desc, input_y.get_data_handle(),
+                            scale_y);
+  }
+
+  inner_product_forward CreateMulPrimitive(const memory &x_memory,
+                                           const memory &y_memory,
+                                           const memory::desc &dst_desc,
+                                           Tensor *output,
+                                           const ExecutionContext &ctx) {
+    const auto y_desc = y_memory.get_primitive_desc().desc();
+    const auto x_desc = x_memory.get_primitive_desc().desc();
+
+    mkldnn::primitive_attr mul_attr;
+
+    bool force_fp32_output = ctx.Attr<bool>("force_fp32_output");
+    auto scale_y_data = ctx.Attr<std::vector<float>>("scale_y");
+    auto scale_x_data = ctx.Attr<float>("scale_x");
+    auto scale_out_data =
+        force_fp32_output ? 1.0f : ctx.Attr<float>("scale_out");
+
+    bool is_multi_channel = scale_y_data.size() > 1;
+    int count = is_multi_channel ? scale_y_data.size() : 1;
+    std::vector<float> output_shift_scale(count);
+    for (int i = 0; i < count; i++) {
+      if (scale_y_data[i] == 0.0)
+        output_shift_scale[i] = scale_out_data;
+      else
+        output_shift_scale[i] =
+            scale_out_data / (scale_x_data * scale_y_data[i]);
+    }
+    int mul_mask = is_multi_channel ? 1 : 0;
+    mul_attr.set_output_scales(mul_mask, output_shift_scale);
```

Comment:
Please consider extracting the creation of `mkldnn::primitive_attr mul_attr` to an external function. It would make `CreateMulPrimitive` much shorter and easier to analyse.

Reply:
yeah...good suggestion, followed.